lib/puppet/util/windows/service.rb

```diff
@@ -0,0 +1,708 @@
+        size_required = 0
+        services_returned = 0
+        FFI::MemoryPointer.new(:dword) do |bytes_pointer|
+          FFI::MemoryPointer.new(:dword) do |svcs_ret_ptr|
+            FFI::MemoryPointer.new(:dword) do |resume_ptr|
+              resume_ptr.write_int(0)
+              # Fetch the bytes of memory required to be allocated
+              # for QueryServiceConfigW to return succesfully. This
+              # is done by sending NULL and 0 for the pointer and size
+              # respectively, letting the command fail, then reading the
+              # value of pcbBytesNeeded
+              #
+              # return value will be false from this call, since it's designed
+              # to fail. Just ignore it
+              EnumServicesStatusExW(
+                scm,
+                SC_ENUM_PROCESS_INFO,
+                ALL_SERVICE_TYPES,
+                SERVICE_STATE_ALL,
+                nil,
+                0,
+                bytes_pointer,
+                svcs_ret_ptr,
+                resume_ptr,
+                nil,
+              )
+              size_required = bytes_pointer.read(:dword)
+              FFI::MemoryPointer.new(size_required) do |buffer_ptr|
+                resume_ptr.write_int(0)
+                svcs_ret_ptr.write_int(0)
+                success = EnumServicesStatusExW(
+                  scm,
+                  SC_ENUM_PROCESS_INFO,
+                  ALL_SERVICE_TYPES,
+                  SERVICE_STATE_ALL,
+                  buffer_ptr,
+                  buffer_ptr.size,
+                  bytes_pointer,
+                  svcs_ret_ptr,
+                  resume_ptr,
+                  nil,
+                )
+                if success == FFI::WIN32_FALSE
+                  raise Puppet::Util::Windows::Error.new(_("Failed to fetch services"))
+                end
+                # Now that the buffer is populated with services
+                # we pull the data from memory using pointer arithmetic:
+                # the number of services returned by the function is
+                # available to be read from svcs_ret_ptr, and we iterate
+                # that many times moving the cursor pointer the length of
+                # ENUM_SERVICE_STATUS_PROCESSW.size. This should iterate
+                # over the buffer and extract each struct.
+                services_returned = svcs_ret_ptr.read(:dword)
+                cursor_ptr = buffer_ptr
+                1.upto(services_returned) do |index|
+                  service = ENUM_SERVICE_STATUS_PROCESSW.new(cursor_ptr)
+                  services[service[:lpServiceName].read_arbitrary_wide_string_up_to(SERVICENAME_MAX)] = {
+                    :display_name => service[:lpDisplayName].read_arbitrary_wide_string_up_to(SERVICENAME_MAX),
+                    :service_status_process => service[:ServiceStatusProcess]
+                  }
+                  cursor_ptr += ENUM_SERVICE_STATUS_PROCESSW.size
```

Comment:
There's a trick you can use to avoid manually incrementing the pointer like this. See https://github.com/puppetlabs/puppet/blob/ba4d1a1aba0095d3c70b98fea5c67434a4876a61/lib/puppet/util/windows/process.rb#L136-L142

Just note that the indexing is 0-based instead of 1-based like you have here